### ISSUE TITLE
[Build] Exclude specific module (r.geomorphon) at compilation

### ISSUE THREAD

**Issue by waarmond** [OSGeo/grass#3731]
## Describe the bug
I can't finish the GRASS build due to one module not compiling:

`Cannot parse interface for module r.geomorphon. Empty strings will be placed instead of description and keywords. Reason: Unable to fetch interface description for command '<r.geomorphon>'.`

This is certainly my fault. Is there any way to disable one specific module at configuration time?

## To reproduce
```
    ./configure \
    --prefix=/opt/$pkgname \
    --with-freetype-includes=/usr/include/freetype2 \
    --with-wxwidgets \
    --with-readline \
    --with-pthread \
    --with-netcdf \
    --with-nls \
    --with-geos \
    --with-postgres \
    --with-pdal \
    --with-bzlib \
    --with-zstd

    --disable-geomorphon ??

  make
```

## System description
- ArchLinux
- git master

**Comment by neteler** [comment]
Please post the full error with surrounding context.

**Comment by waarmond** [comment]
Is LC_ALL=C necessary?
```
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu“ wird ignoriert
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/bin.x86_64-pc-linux-gnu“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/bin.x86_64-pc-linux-gnu“ wird ignoriert
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/include/grass“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/include/grass“ wird ignoriert
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/lib“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/lib“ wird ignoriert
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/bin“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/bin“ wird ignoriert
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/etc“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/etc“ wird ignoriert
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/driver“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/driver“ wird ignoriert
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/driver/db“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/driver/db“ wird ignoriert
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/fonts“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/fonts“ wird ignoriert
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/docs“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/docs“ wird ignoriert
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/docs/html“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/docs/html“ wird ignoriert
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/docs/man“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/docs/man“ wird ignoriert
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/docs/man/man1“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/docs/man/man1“ wird ignoriert
../include/Make/Rules.make:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/utils“ wird überschrieben
../include/Make/Rules.make:16: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/utils“ wird ignoriert
../include/Make/Rules.make:19: Warnung: Das Rezept für das Ziel „OBJ.x86_64-pc-linux-gnu“ wird überschrieben
../include/Make/Rules.make:19: Warnung: Das alte Rezept für das Ziel „OBJ.x86_64-pc-linux-gnu“ wird ignoriert
../include/Make/Rules.make:56: Warnung: Das Rezept für das Ziel „clean“ wird überschrieben
../include/Make/Rules.make:56: Warnung: Das alte Rezept für das Ziel „clean“ wird ignoriert
WARNING: Weisberg-Bingham's W'' cannot be used for n < 50 or n > 99
make[4]: *** [../../include/Make/Html.make:14: r.geomorphon.tmp.html] Fehler 1
Makefile:13: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/etc/d.polar“ wird überschrieben
../../include/Make/ScriptRules.make:19: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/etc/d.polar“ wird ignoriert
Makefile:16: Warnung: Das Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/etc/db.test“ wird überschrieben
../../include/Make/ScriptRules.make:19: Warnung: Das alte Rezept für das Ziel „./src/grass/dist.x86_64-pc-linux-gnu/etc/db.test“ wird ignoriert
Creating translations (= 'make mo')
3361 übersetzte Meldungen, 164 ungenaue Übersetzungen, 747 unübersetzte Meldungen.
0 übersetzte Meldungen, 4272 unübersetzte Meldungen.
646 übersetzte Meldungen, 86 ungenaue Übersetzungen, 3540 unübersetzte Meldungen.
2192 übersetzte Meldungen, 142 ungenaue Übersetzungen, 1938 unübersetzte Meldungen.
po/grasswxpy_de.po:20658: Mehrfachdefinition der Meldung...
po/grasswxpy_de.po:4581: ...dies ist die Stelle der ersten Definition
msgfmt: found 1 fatal error
make[3]: *** [Makefile:100: ./src/grass/dist.x86_64-pc-linux-gnu/locale/de/LC_MESSAGES/grasswxpy.mo] Fehler 1
```

**Comment by wenzeslaus** [comment]
The "Cannot parse interface for module r.geomorphon." message is not a fatal error. GRASS GIS should compile regardless of these messages. With that said, the message is likely an indication of another error. r.geomorphon is one of the tools which should compile in any case (hence there is no flag to disable that one). Aren't there other messages but for other tools? 

Is there a list of directories at the end of the compilation output with errors? It looks like this:

```
make[1]: Nothing to be done for 'manifests'.
make[1]: Leaving directory '/home/vpetras/Projects/grass/code/grass'
GRASS GIS 8.4.0dev 4f0137d0fb compilation log
--------------------------------------------------
Started compilation: Thu 23 May 2024 09:11:04 PM EDT
--
Errors in:
/home/vpetras/Projects/grass/code/grass/scripts/g.extension
--
In case of errors please change into the directory with error and run 'make'.
If you get multiple errors, you need to deal with them in the order they
appear in the error log. If you get an error building a library, you will
also get errors from anything which uses the library.
--
Finished compilation: Thu 23 May 2024 09:11:39 PM EDT
```

**Comment by neteler** [comment]
There is apparently a duplicate message after yesterday's gettext update in #3729:

```
po/grasswxpy_de.po:20658: Mehrfachdefinition der Meldung...
po/grasswxpy_de.po:4581: ...dies ist die Stelle der ersten Definition
msgfmt
```

We need to deduplicate the file in a new PR (and optimize the pot file update procedure).

**Comment by neteler** [comment]
See also #3465

**Comment by waarmond** [comment]
`[./r.geomorphon]$ LC_ALL=C make`

```
if [ "./grass/dist.x86_64-pc-linux-gnu/bin/r.geomorphon" != "" ] ; then GISRC=./grass/dist.x86_64-pc-linux-gnu/demolocation/.grassrc84 GISBASE=./grass/dist.x86_64-pc-linux-gnu PATH="./grass/dist.x86_64-pc-linux-gnu/bin:./grass/dist.x86_64-pc-linux-gnu/bin:./grass/dist.x86_64-pc-linux-gnu/scripts:$PATH" PYTHONPATH="./grass/dist.x86_64-pc-linux-gnu/etc/python:./grass/dist.x86_64-pc-linux-gnu/gui/wxpython:$PYTHONPATH" LD_LIBRARY_PATH="./grass/dist.x86_64-pc-linux-gnu/bin:./grass/dist.x86_64-pc-linux-gnu/bin:./grass/dist.x86_64-pc-linux-gnu/scripts:./grass/dist.x86_64-pc-linux-gnu/lib:./grass/dist.x86_64-pc-linux-gnu/lib:/usr/lib/dri/:" LC_ALL=C LANG=C LANGUAGE=C ./grass/dist.x86_64-pc-linux-gnu/bin/r.geomorphon --html-description < /dev/null | grep -v '</body>\|</html>\|</div> <!-- end container -->' > r.geomorphon.tmp.html ; fi
make: *** [../../include/Make/Html.make:14: r.geomorphon.tmp.html] Error 1
rm r.geomorphon.tmp.html
```

**Comment by waarmond** [comment]
I'm on Python 3.10, but `r.geomorphon` tries to load 3.11:
`gdb: error while loading shared libraries: libpython3.11.so.1.0:`

**Comment by echoix** [comment]
Just before continuing further, was the whole grass all configured+compiled fresh, or its an incremental compilation with existing libraries or installation on the system?

**Comment by waarmond** [comment]
I've deinstalled my existing grass installation. The build is happening through an ArchLinux PKGBUILD with a clean /src folder.

**Comment by neteler** [comment]
@waarmond reopening in case there are other issues.

**Comment by waarmond** [comment]
Thanks for reopening. Sadly I won't be able to recompile before tomorrow.

**Comment by waarmond** [comment]
Here is the strace of my r.geomorphon crash:
```
openat(AT_FDCWD, "/usr/etc/ucx/ucx.conf", O_RDONLY) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_ACCERR, si_addr=0x55ec34d73ea0} ---
```
Does anybody know openucx from OpenMPI ?

**Comment by wenzeslaus** [comment]
These error don't seem to be related to GRASS GIS. There is no OpenMPI or MPI specific code in this repo. Some very specific local system configuration or misconfiguration?

**Comment by marisn** [comment]
> Here is the strace of my r.geomorphon crash:

This is a red herring. Run under GDB to get a backtrace of crash or under valgrind if memory corruption is suspected based on the GDB backtrace.
Do you use any exotic compiler or flags (e.g. -O3)? Although recently we have done a great job to make code base not so GCC specific and cleaned up a lot of warnings, I am certain there is still a lot to do.

**Comment by waarmond** [comment]
Is r.geomorphon using GDAL?

Via gdal → armadillo → arpack → OpenMPI it was broken for me.

**Comment by neteler** [comment]
Concerning GDAL:

> Now I'm getting to the Pudel's Kern: `./src/grass/dist.x86_64-pc-linux-gnu/bin/r.geomorphon: error while loading shared libraries: libgrass_raster.8.4.so: cannot open shared object file:`
> 
> which exists at ./src/grass/dist.x86_64-pc-linux-gnu/lib/libgrass_raster.8.4.so

What does `ldd` report on the binary?

**Comment by parona-source** [comment]
On Gentoo this issue appears to have come up after [the move to flexiblas](https://labs.quansight.org/blog/blas-lapack-packaging). Im torn on claiming if its the same issue exactly as I don't know if or when Arch wouldve used flexiblas, but the snippet of the gdb output with it failing in dirname makes me think these are the same.

https://bugs.gentoo.org/965293#c11

```
if [ "/var/tmp/portage/sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/bin/r.geomorphon" != "" ] ; then GISRC=/var/tmp/portage/sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/demoloc
ation/.grassrc84 GISBASE=/var/tmp/portage/sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu PATH="/var/tmp/portage/sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/bin:/var/tmp/portage/
sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/bin:/var/tmp/portage/sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/scripts:$PATH" PYTHONPATH="/var/tmp/portage/sci-geosciences/grass
-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/etc/python:/var/tmp/portage/sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/gui/wxpython:$PYTHONPATH" LD_LIBRARY_PATH="/var/tmp/portage/sci-geosciences/gra
ss-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/bin:/var/tmp/portage/sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/bin:/var/tmp/portage/sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-
linux-gnu/scripts:/var/tmp/portage/sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/lib:/var/tmp/portage/sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/lib:" LC_ALL=C LANG=C LANGUAGE
=C /var/tmp/portage/sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/bin/r.geomorphon --html-description < /dev/null | grep -v '</body>\|</html>\|</div> <!-- end container -->' > r.geomorphon.tmp.html ; fi
```

```
Reading symbols from /var/tmp/portage/sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/bin/r.geomorphon...
[New LWP 8245]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `/var/tmp/portage/sci-geosciences/grass-8.4.1-r1/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/bin/r.geomorphon --html-description'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000055675ac51200 in dirname ()
(gdb) bt
#0  0x000055675ac51200 in dirname ()
#1  0x00007fd9c212d126 in __flexiblas_init_default_paths ()
    at /usr/src/debug/sci-libs/flexiblas-3.5.0/flexiblas-release-v3.5.0/src/paths.c:152
#2  0x00007fd9c2115cca in flexiblas_init ()
    at /usr/src/debug/sci-libs/flexiblas-3.5.0/flexiblas-release-v3.5.0/src/flexiblas.c:683
#3  0x00007fd9d1bf3c3e in call_init (l=<optimized out>, argc=argc@entry=2, argv=argv@entry=0x7fff077750d8, 
    env=env@entry=0x7fff077750f0) at dl-init.c:74
#4  0x00007fd9d1bf3d43 in call_init (l=<optimized out>, argc=<optimized out>, argv=<optimized out>, 
    env=<optimized out>) at dl-init.c:120
#5  _dl_init (main_map=0x7fd9d1c2c2f0, argc=2, argv=0x7fff077750d8, env=0x7fff077750f0) at dl-init.c:121
#6  0x00007fd9d1c11220 in _dl_start_user () from /lib64/ld-linux-x86-64.so.2
#7  0x0000000000000002 in ?? ()
#8  0x00007fff07777953 in ?? ()
#9  0x00007fff077779be in ?? ()
#10 0x0000000000000000 in ?? ()
(gdb) f 1
#1  0x00007fd9c212d126 in __flexiblas_init_default_paths ()
    at /usr/src/debug/sci-libs/flexiblas-3.5.0/flexiblas-release-v3.5.0/src/paths.c:152
152             char *folder = dirname(libpath);
(gdb) p libpath
$1 = 0x55676de36ee0 ""
```

https://github.com/mpimd-csc/flexiblas/blob/v3.5.0/src/paths.c#L145
https://github.com/mpimd-csc/flexiblas/blob/v3.5.0/src/paths.c#L60

To me this seems like a flexiblas bug. Looks like for me its coming from gdal -> armadillo -> lapack(flexiblas). Disabling armadillo support in gdal gets around the issue.

<details>
<summary>Lddtree of r.geomorphon</summary>

```
LD_LIBRARY_PATH="dist.x86_64-pc-linux-gnu/lib" lddtree dist.x86_64-pc-linux-gnu/bin/r.geomorphon 
dist.x86_64-pc-linux-gnu/bin/r.geomorphon (interpreter => /lib64/ld-linux-x86-64.so.2)
    libgrass_raster.8.4.so => /var/tmp/portage/sci-geosciences/grass-8.4.1-r2/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/lib/libgrass_raster.8.4.so
        libgrass_gproj.8.4.so => /var/tmp/portage/sci-geosciences/grass-8.4.1-r2/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/lib/libgrass_gproj.8.4.so
            libproj.so.25 => /usr/lib64/libproj.so.25
                libsqlite3.so.0 => /usr/lib64/libsqlite3.so.0
                    libz.so.1 => /usr/lib64/libz.so.1
                    libicuio.so.77 => /usr/lib64/libicuio.so.77
                    libicui18n.so.77 => /usr/lib64/libicui18n.so.77
                    libicuuc.so.77 => /usr/lib64/libicuuc.so.77
                        libicudata.so.77 => /usr/lib64/libicudata.so.77
                libtiff.so.6 => /usr/lib64/libtiff.so.6
                    libwebp.so.7 => /usr/lib64/libwebp.so.7
                        libsharpyuv.so.0 => /usr/lib64/libsharpyuv.so.0
                    libzstd.so.1 => /usr/lib64/libzstd.so.1
                    liblzma.so.5 => /usr/lib64/liblzma.so.5
                    libjpeg.so.62 => /usr/lib64/libjpeg.so.62
                libstdc++.so.6 => /usr/lib/gcc/x86_64-pc-linux-gnu/16/libstdc++.so.6
                libgcc_s.so.1 => /usr/lib/gcc/x86_64-pc-linux-gnu/16/libgcc_s.so.1
        libgdal.so.38 => /usr/lib64/libgdal.so.38
            libarchive.so.13 => /usr/lib64/libarchive.so.13
                libacl.so.1 => /usr/lib64/libacl.so.1
                libbz2.so.1 => /usr/lib64/libbz2.so.1
            libcurl.so.4 => /usr/lib64/libcurl.so.4
                libcares.so.2 => /usr/lib64/libcares.so.2
                libnghttp3.so.9 => /usr/lib64/libnghttp3.so.9
                libnghttp2.so.14 => /usr/lib64/libnghttp2.so.14
                libpsl.so.5 => /usr/lib64/libpsl.so.5
                    libidn2.so.0 => /usr/lib64/libidn2.so.0
                    libunistring.so.5 => /usr/lib64/libunistring.so.5
                libssl.so.3 => /usr/lib64/libssl.so.3
            libodbc.so.2 => /usr/lib64/libodbc.so.2
                libltdl.so.7 => /usr/lib64/libltdl.so.7
            libodbcinst.so.2 => /usr/lib64/libodbcinst.so.2
            libxml2.so.16 => /usr/lib64/libxml2.so.16
            libcryptopp.so.8 => /usr/lib64/libcryptopp.so.8
                libgomp.so.1 => /usr/lib/gcc/x86_64-pc-linux-gnu/16/libgomp.so.1
            libcrypto.so.3 => /usr/lib64/libcrypto.so.3
            libdeflate.so.0 => /usr/lib64/libdeflate.so.0
            liblz4.so.1 => /usr/lib64/liblz4.so.1
            libblosc.so.1 => /usr/lib64/libblosc.so.1
                libsnappy.so.1 => /usr/lib64/libsnappy.so.1
            libarmadillo.so.15 => /usr/lib64/libarmadillo.so.15
                liblapack.so.3 => /usr/lib64/liblapack.so.3
            libqhull_r.so.8.1 => /usr/lib64/libqhull_r.so.8.1
            libxerces-c-3.3.so => /usr/lib64/libxerces-c-3.3.so
            libmuparser.so.2 => /usr/lib64/libmuparser.so.2
            libLerc.so.4 => /usr/lib64/libLerc.so.4
            libgeotiff.so.5 => /usr/lib64/libgeotiff.so.5
            libavif.so.16 => /usr/lib64/libavif.so.16
                libaom.so.3 => /usr/lib64/libaom.so.3
            libheif.so.1 => /usr/lib64/libheif.so.1
                libde265.so.0 => /usr/lib64/libde265.so.0
                libkvazaar.so.7 => /usr/lib64/libkvazaar.so.7
            libpng16.so.16 => /usr/lib64/libpng16.so.16
            libaec.so.0 => /usr/lib64/libaec.so.0
            libpoppler.so.154 => /usr/lib64/libpoppler.so.154
                libfreetype.so.6 => /usr/lib64/libfreetype.so.6
                    libbrotlidec.so.1 => /usr/lib64/libbrotlidec.so.1
                        libbrotlicommon.so.1 => /usr/lib64/libbrotlicommon.so.1
                libfontconfig.so.1 => /usr/lib64/libfontconfig.so.1
                liblcms2.so.2 => /usr/lib64/liblcms2.so.2
                libsmime3.so => /usr/lib64/libsmime3.so
                    libnssutil3.so => /usr/lib64/libnssutil3.so
                    libplds4.so => /usr/lib64/libplds4.so
                libnss3.so => /usr/lib64/libnss3.so
                libplc4.so => /usr/lib64/libplc4.so
                libnspr4.so => /usr/lib64/libnspr4.so
            libgif.so.7 => /usr/lib64/libgif.so.7
            libnetcdf.so.22 => /usr/lib64/libnetcdf.so.22
                libhdf5_hl.so.310 => /usr/lib64/libhdf5_hl.so.310
                libzip.so.5 => /usr/lib64/libzip.so.5
            libcfitsio.so.10 => /usr/lib64/libcfitsio.so.10
            libhdf5.so.310 => /usr/lib64/libhdf5.so.310
                libsz.so.2 => /usr/lib64/libsz.so.2
            libpq.so.5 => /usr/lib64/libpq.so.5
                libgssapi_krb5.so.2 => /usr/lib64/libgssapi_krb5.so.2
                    libkrb5.so.3 => /usr/lib64/libkrb5.so.3
                    libk5crypto.so.3 => /usr/lib64/libk5crypto.so.3
                    libcom_err.so.2 => /usr/lib64/libcom_err.so.2
                    libkrb5support.so.0 => /usr/lib64/libkrb5support.so.0
                    libkeyutils.so.1 => /usr/lib64/libkeyutils.so.1
                    libresolv.so.2 => /lib64/libresolv.so.2
            libopenjp2.so.7 => /usr/lib64/libopenjp2.so.7
            libOpenEXR-3_3.so.32 => /usr/lib64/libOpenEXR-3_3.so.32
                libIlmThread-3_3.so.32 => /usr/lib64/libIlmThread-3_3.so.32
                libOpenEXRCore-3_3.so.32 => /usr/lib64/libOpenEXRCore-3_3.so.32
                libIex-3_3.so.32 => /usr/lib64/libIex-3_3.so.32
            libclntsh.so.21.1 => /usr/lib64/oracle/client/lib/libclntsh.so.21.1
                libnnz21.so => /usr/lib64/oracle/client/lib/libnnz21.so
                libdl.so.2 => /lib64/libdl.so.2
                libpthread.so.0 => /lib64/libpthread.so.0
                librt.so.1 => /lib64/librt.so.1
                libaio.so.1 => /usr/lib64/libaio.so.1
                libclntshcore.so.21.1 => /usr/lib64/oracle/client/lib/libclntshcore.so.21.1
            libpcre2-8.so.0 => /usr/lib64/libpcre2-8.so.0
            libspatialite.so.8 => /usr/lib64/libspatialite.so.8
                libminizip.so.1 => /usr/lib64/libminizip.so.1
            libmysqlclient.so.21 => /usr/lib64/libmysqlclient.so.21
            libfreexl.so.1 => /usr/lib64/libfreexl.so.1
            libarrow_dataset.so.2200 => /usr/lib64/libarrow_dataset.so.2200
                libarrow_acero.so.2200 => /usr/lib64/libarrow_acero.so.2200
            libexpat.so.1 => /usr/lib64/libexpat.so.1
            libgeos_c.so.1 => /usr/lib64/libgeos_c.so.1
                libgeos.so.3.14.0 => /usr/lib64/libgeos.so.3.14.0
            libjson-c.so.5 => /usr/lib64/libjson-c.so.5
            libparquet.so.2200 => /usr/lib64/libparquet.so.2200
                libthrift.so.0.21.0 => /usr/lib64/libthrift.so.0.21.0
            libarrow_compute.so.2200 => /usr/lib64/libarrow_compute.so.2200
                libre2.so.11 => /usr/lib64/libre2.so.11
                    libabsl_log_internal_check_op.so.2508.0.0 => /usr/lib64/libabsl_log_internal_check_op.so.2508.0.0
                        libabsl_leak_check.so.2508.0.0 => /usr/lib64/libabsl_leak_check.so.2508.0.0
                        libabsl_log_internal_nullguard.so.2508.0.0 => /usr/lib64/libabsl_log_internal_nullguard.so.2508.0.0
                    libabsl_log_internal_message.so.2508.0.0 => /usr/lib64/libabsl_log_internal_message.so.2508.0.0
                        libabsl_examine_stack.so.2508.0.0 => /usr/lib64/libabsl_examine_stack.so.2508.0.0
                            libabsl_stacktrace.so.2508.0.0 => /usr/lib64/libabsl_stacktrace.so.2508.0.0
                                libabsl_debugging_internal.so.2508.0.0 => /usr/lib64/libabsl_debugging_internal.so.2508.0.0
                                libabsl_malloc_internal.so.2508.0.0 => /usr/lib64/libabsl_malloc_internal.so.2508.0.0
                            libabsl_symbolize.so.2508.0.0 => /usr/lib64/libabsl_symbolize.so.2508.0.0
                                libabsl_demangle_internal.so.2508.0.0 => /usr/lib64/libabsl_demangle_internal.so.2508.0.0
                                    libabsl_demangle_rust.so.2508.0.0 => /usr/lib64/libabsl_demangle_rust.so.2508.0.0
                                        libabsl_decode_rust_punycode.so.2508.0.0 => /usr/lib64/libabsl_decode_rust_punycode.so.2508.0.0
                                            libabsl_utf8_for_code_point.so.2508.0.0 => /usr/lib64/libabsl_utf8_for_code_point.so.2508.0.0
                        libabsl_log_internal_format.so.2508.0.0 => /usr/lib64/libabsl_log_internal_format.so.2508.0.0
                        libabsl_log_internal_structured_proto.so.2508.0.0 => /usr/lib64/libabsl_log_internal_structured_proto.so.2508.0.0
                        libabsl_strerror.so.2508.0.0 => /usr/lib64/libabsl_strerror.so.2508.0.0
                        libabsl_log_internal_log_sink_set.so.2508.0.0 => /usr/lib64/libabsl_log_internal_log_sink_set.so.2508.0.0
                            libabsl_log_sink.so.2508.0.0 => /usr/lib64/libabsl_log_sink.so.2508.0.0
                        libabsl_log_internal_globals.so.2508.0.0 => /usr/lib64/libabsl_log_internal_globals.so.2508.0.0
                        libabsl_log_globals.so.2508.0.0 => /usr/lib64/libabsl_log_globals.so.2508.0.0
                        libabsl_log_internal_proto.so.2508.0.0 => /usr/lib64/libabsl_log_internal_proto.so.2508.0.0
                        libabsl_time.so.2508.0.0 => /usr/lib64/libabsl_time.so.2508.0.0
                            libabsl_time_zone.so.2508.0.0 => /usr/lib64/libabsl_time_zone.so.2508.0.0
                        libabsl_strings_internal.so.2508.0.0 => /usr/lib64/libabsl_strings_internal.so.2508.0.0
                        libabsl_base.so.2508.0.0 => /usr/lib64/libabsl_base.so.2508.0.0
                    libabsl_raw_hash_set.so.2508.0.0 => /usr/lib64/libabsl_raw_hash_set.so.2508.0.0
                        libabsl_hashtablez_sampler.so.2508.0.0 => /usr/lib64/libabsl_hashtablez_sampler.so.2508.0.0
                    libabsl_str_format_internal.so.2508.0.0 => /usr/lib64/libabsl_str_format_internal.so.2508.0.0
                        libabsl_int128.so.2508.0.0 => /usr/lib64/libabsl_int128.so.2508.0.0
                    libabsl_hash.so.2508.0.0 => /usr/lib64/libabsl_hash.so.2508.0.0
                        libabsl_city.so.2508.0.0 => /usr/lib64/libabsl_city.so.2508.0.0
                    libabsl_synchronization.so.2508.0.0 => /usr/lib64/libabsl_synchronization.so.2508.0.0
                        libabsl_kernel_timeout_internal.so.2508.0.0 => /usr/lib64/libabsl_kernel_timeout_internal.so.2508.0.0
                        libabsl_tracing_internal.so.2508.0.0 => /usr/lib64/libabsl_tracing_internal.so.2508.0.0
                    libabsl_strings.so.2508.0.0 => /usr/lib64/libabsl_strings.so.2508.0.0
                    libabsl_raw_logging_internal.so.2508.0.0 => /usr/lib64/libabsl_raw_logging_internal.so.2508.0.0
                    libabsl_spinlock_wait.so.2508.0.0 => /usr/lib64/libabsl_spinlock_wait.so.2508.0.0
                libutf8proc.so.3 => /usr/lib64/libutf8proc.so.3
            libarrow.so.2200 => /usr/lib64/libarrow.so.2200
                libbrotlienc.so.1 => /usr/lib64/libbrotlienc.so.1
    libgrass_gis.8.4.so => /var/tmp/portage/sci-geosciences/grass-8.4.1-r2/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/lib/libgrass_gis.8.4.so
        libgrass_datetime.8.4.so => /var/tmp/portage/sci-geosciences/grass-8.4.1-r2/work/grass-8.4.1/dist.x86_64-pc-linux-gnu/lib/libgrass_datetime.8.4.so
    libm.so.6 => /lib64/libm.so.6
    libc.so.6 => /lib64/libc.so.6
```
</details>


Edit: lol, even lmao

```
$ GISRC=/home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/demolocation/.grassrc85 GISBASE=/home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu PATH="/home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/bin:/home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/bin:/home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/scripts:$PATH" PYTHONPATH="/home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/etc/python:/home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/gui/wxpython:$PYTHONPATH" LD_LIBRARY_PATH="/home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/bin:/home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/bin:/home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/scripts:/home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/lib:/home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/lib:" LC_ALL=C LANG=C LANGUAGE=C gdb --args /home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/bin/r.geomorphon --html-description
GNU gdb (Gentoo 16.3 vanilla) 16.3
Copyright (C) 2024 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://bugs.gentoo.org/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/bin/r.geomorphon...
(gdb) b /usr/src/debug/sci-libs/flexiblas-3.5.0/flexiblas-release-v3.5.0/src/paths.c:89
No source file named /usr/src/debug/sci-libs/flexiblas-3.5.0/flexiblas-release-v3.5.0/src/paths.c.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (/usr/src/debug/sci-libs/flexiblas-3.5.0/flexiblas-release-v3.5.0/src/paths.c:89) pending.
(gdb) r
Starting program: /home/ask/sources/archive/grass/dist.x86_64-pc-linux-gnu/bin/r.geomorphon --html-description
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Breakpoint 1, __flexiblas_get_library_location_impl ()
    at /usr/src/debug/sci-libs/flexiblas-3.5.0/flexiblas-release-v3.5.0/src/paths.c:89
89          char * buffer = malloc(sizeof(char) * path_max);
(gdb) n
95              if (dladdr(FUNC_RETURN_ADDRESS(), &info))
(gdb) n
97                  if (!realpath(info.dli_fname, buffer)) {break;}
(gdb) n
104         return buffer;
(gdb) n
__flexiblas_init_default_paths ()
    at /usr/src/debug/sci-libs/flexiblas-3.5.0/flexiblas-release-v3.5.0/src/paths.c:150
150         if (libpath != NULL)
(gdb) p libpath
$1 = 0x555555faeee0 "/usr/lib64/libflexiblas.so.3.5"
(gdb) s
152             char *folder = dirname(libpath);
(gdb) p *dirname
$2 = 0x55555555d9bb "NE"
```

https://github.com/OSGeo/grass/blob/8.4.1/raster/r.geomorphon/pattern.c#L9

**Comment by parona-source** [comment]
So the issue is that grass redefines the glibc function dirname to an char* array. Flexiblas then tries to execute it.

In this light its possible that the original bug didn't come from flexiblas, just a similar victim.

https://sourceware.org/glibc/manual/latest/html_node/Reserved-Names.html

Thank you @thesamesam for the pointers.

**Comment by nilason** [comment]
> So the issue is that grass redefines the glibc function dirname to an char* array. Flexiblas then tries to execute it.
> 
> In this light its possible that the original bug didn't come from flexiblas, just a similar victim.
> 
> https://sourceware.org/glibc/manual/latest/html_node/Reserved-Names.html
> 
> Thank you [@thesamesam](https://github.com/thesamesam) for the pointers.

If I understand you correctly, changing the name of this variable:

https://github.com/OSGeo/grass/blob/3999ef44f69b088db09873848c8afd13931d98c5/raster/r.geomorphon/local_proto.h#L89

would solve this issue?

**Comment by parona-source** [comment]
> If I understand you correctly, changing the name of this variable:
> 
> [grass/raster/r.geomorphon/local_proto.h](https://github.com/OSGeo/grass/blob/3999ef44f69b088db09873848c8afd13931d98c5/raster/r.geomorphon/local_proto.h#L89)
> 
> Line 89 in [3999ef4](/OSGeo/grass/commit/3999ef44f69b088db09873848c8afd13931d98c5)
>  extern const char *dirname[]; 
> 
> would solve this issue?

I think so, currently trying it.

Edit:
```
GRASS 8.5.0dev 06ccd10965 compilation log
--------------------------------------------------
Started compilation: Mon 17 Nov 08:55:50 EET 2025
--
Errors in:
No errors detected.
--
Finished compilation: Mon 17 Nov 09:01:06 EET 2025
```

🥳 

<details>
<summary>diff, didnt think too hard on the name</summary>

```diff
diff --git a/raster/r.geomorphon/local_proto.h b/raster/r.geomorphon/local_proto.h
index ff5de55db5..56fae9c5d6 100644
--- a/raster/r.geomorphon/local_proto.h
+++ b/raster/r.geomorphon/local_proto.h
@@ -86,7 +86,7 @@ const char *form_long_name(const FORMS);
 
 /* pattern */
 int calc_pattern(PATTERN *pattern, int row, int cur_row, int col, const int);
-extern const char *dirname[];
+extern const char *direction_dirname[];
 
 /* geom */
 void generate_ternary_codes(void);
diff --git a/raster/r.geomorphon/pattern.c b/raster/r.geomorphon/pattern.c
index 59f457a7b8..0fa1e8b321 100644
--- a/raster/r.geomorphon/pattern.c
+++ b/raster/r.geomorphon/pattern.c
@@ -6,7 +6,7 @@
  * 5|6|7 */
 static int nextr[NUM_DIRS] = {-1, -1, -1, 0, 1, 1, 1, 0};
 static int nextc[NUM_DIRS] = {1, 0, -1, -1, -1, 0, 1, 1};
-const char *dirname[NUM_DIRS] = {"NE", "N", "NW", "W", "SW", "S", "SE", "E"};
+const char *direction_dirname[NUM_DIRS] = {"NE", "N", "NW", "W", "SW", "S", "SE", "E"};
 
 /*
  * A more thorough comparison using a few factors of different priority
@@ -125,7 +125,7 @@ int calc_pattern(PATTERN *pattern, int row, int cur_row, int col,
             zenith_easting = nadir_easting = target_easting;
             pattern->e[i] = cur_easting;
             pattern->n[i] = cur_northing;
-            prof_sso(dirname[i]);
+            prof_sso(direction_dirname[i]);
         }
         while (cur_distance < search_distance) {
             if (cur_row + j * nextr[i] < 0 ||
diff --git a/raster/r.geomorphon/profile.c b/raster/r.geomorphon/profile.c
index 304250954e..7fcc8e36a9 100644
--- a/raster/r.geomorphon/profile.c
+++ b/raster/r.geomorphon/profile.c
@@ -145,42 +145,42 @@ void prof_pattern(const double o_elevation, const PATTERN *p)
 
     prof_sso("pattern");
     for (i = 0; i < NUM_DIRS; i++)
-        prof_int(dirname[i], p->pattern[i]);
+        prof_int(direction_dirname[i], p->pattern[i]);
     prof_eso();
 
     prof_sso("rel_elevation_m");
     for (i = 0; i < NUM_DIRS; i++)
-        prof_mtr(dirname[i], p->elevation[i]);
+        prof_mtr(direction_dirname[i], p->elevation[i]);
     prof_eso();
 
     prof_sso("abs_elevation_m");
     for (i = 0; i < NUM_DIRS; i++)
-        prof_mtr(dirname[i], o_elevation + p->elevation[i]);
+        prof_mtr(direction_dirname[i], o_elevation + p->elevation[i]);
     prof_eso();
 
     prof_sso("distance_m");
     for (i = 0; i < NUM_DIRS; i++)
-        prof_mtr(dirname[i], p->distance[i]);
+        prof_mtr(direction_dirname[i], p->distance[i]);
     prof_eso();
 
     prof_sso("offset_easting_m");
     for (i = 0; i < NUM_DIRS; i++)
-        prof_mtr(dirname[i], p->x[i]);
+        prof_mtr(direction_dirname[i], p->x[i]);
     prof_eso();
 
     prof_sso("offset_northing_m");
     for (i = 0; i < NUM_DIRS; i++)
-        prof_mtr(dirname[i], p->y[i]);
+        prof_mtr(direction_dirname[i], p->y[i]);
     prof_eso();
 
     prof_sso("easting");
     for (i = 0; i < NUM_DIRS; i++)
-        prof_dbl(dirname[i], p->e[i]);
+        prof_dbl(direction_dirname[i], p->e[i]);
     prof_eso();
 
     prof_sso("northing");
     for (i = 0; i < NUM_DIRS; i++)
-        prof_dbl(dirname[i], p->n[i]);
+        prof_dbl(direction_dirname[i], p->n[i]);
     prof_eso();
 }
 ```

</details>

**Comment by nilason** [comment]
Either way, we should rename it anyway.

**Comment by nilason** [comment]
@parona-source Thanks for that great debugging effort!
